mixnet/node.py

```diff
@@ -1,107 +1,186 @@
+            )
+        )
+        peer.mixgossip_channel.add_conn(
+            DuplexConnection(
+                outbound_conn,
+                MixOutboundConnection(
+                    inbound_conn, self.global_config.transmission_rate_per_sec
+                ),
+            )
+        )
+
+    async def send_message(self, msg: bytes):
+        for packet, _ in PacketBuilder.build_real_packets(
+            msg, self.global_config.membership
+        ):
+            await self.mixgossip_channel.gossip(packet)
 
-    async def __run(self):
-        """
-        Read SphinxPackets from inbound socket and spawn a thread for each packet to process it.
 
-        This thread approximates a M/M/inf queue.
-        """
+class MixGossipChannel:
+    peering_degree: int
+    conns: list[DuplexConnection]
+    handler: Callable[[SphinxPacket], Awaitable[NetworkPacket | None]]
+    msg_cache: set[NetworkPacket]
 
+    def __init__(
+        self,
+        peer_degree: int,
+        handler: Callable[[SphinxPacket], Awaitable[NetworkPacket | None]],
+    ):
+        self.peering_degree = peer_degree
+        self.conns = []
+        self.handler = handler
+        self.msg_cache = set()
         # A set just for gathering a reference of tasks to prevent them from being garbage collected.
         # https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task
         self.tasks = set()
 
+    def add_conn(self, conn: DuplexConnection):
+        if len(self.conns) >= self.peering_degree:
+            # For simplicity of the spec, reject the connection if the peering degree is reached.
+            raise ValueError("The peering degree is reached.")
+
+        self.conns.append(conn)
+        task = asyncio.create_task(self.__process_inbound_conn(conn))
+        self.tasks.add(task)
+        # To discard the task from the set automatically when it is done.
+        task.add_done_callback(self.tasks.discard)
+
+    async def __process_inbound_conn(self, conn: DuplexConnection):
         while True:
-            _, packet = await self.inbound_socket.get()
-            task = asyncio.create_task(
-                self.__process_packet(
-                    packet,
-                    self.config.encryption_private_key,
-                    self.config.delay_rate_per_min,
-                )
-            )
-            self.tasks.add(task)
-            # To discard the task from the set automatically when it is done.
-            task.add_done_callback(self.tasks.discard)
+            elem = await conn.recv()
+            # In practice, data transmitted through connections is going to be always 'bytes'.
+            # But here, we use the SphinxPacket type explicitly for simplicity
+            # without implementing serde for SphinxPacket.
+            if isinstance(elem, bytes):
+                assert elem == build_noise_packet()
+                # Drop packet
+                continue
+            elif isinstance(elem, SphinxPacket):
```

Comment:
Nitpick:  You can probably [pattern-match](https://peps.python.org/pep-0636/) here

Reply:
4aa1ee0 Now that I implemented serde for SphinxPacket in `pysphinx` package, I simplified this code using pattern matching.